### DISCUSSION TITLE
support recursive with dominate.document()

### DISCUSSION THREAD

**Discussion by MarisaKirisame** [Knio/dominate#163] [Q&A] [answered]
```
import dominate
from dominate.tags import *

with dominate.document() as outer_doc:
    with dominate.document() as inner_doc:
        p('meow')
    print(inner_doc)
print(outer_doc)
```
output
```
<!DOCTYPE html>
<html>
  <head>
    <title>Dominate</title>
  </head>
  <body>
    <p>meow</p>
  </body>
</html>
<!DOCTYPE html>
<html>
  <head>
    <title>Dominate</title>
  </head>
  <body>
    <html>
      <head>
        <title>Dominate</title>
      </head>
      <body>
        <p>meow</p>
      </body>
    </html>
  </body>
</html>
```

**Comment by Knio** [comment]
No issue/problem/question stated and this looks working as intended

**Comment by MarisaKirisame** [comment]
@Knio 
The <html> tag represents the root of an HTML document. (from https://www.w3schools.com/tags/tag_html.asp)
so - i dont think it should nest, which is what is happening if you look at the output from the second occurence of <!DOCTYPE ...>
instead, I think a better design choice is, when a html root is entered/exited, it does not register itself as a child of what is on the context, breaking the chain.
roughly what happend is that 
`with dominate.document() as xxx: ...`
should has no other effect then writing xxx.

This come up in real life, when I try to use dominate, as I try to create multiple webpage in a hirearchical order. I initially wrote a recursive function to do that, starting from the main entry, generating hyperlink on needed, but as dominate doesnt support that I have to rewrite it from a top-down to a bottom-up manner.

**Comment by Knio** [comment, marked as answer]
> i dont think it should nest

This project is not meant to be an html validator - i.e. it doesn't check that `<td>` is inside a `<tr>`, or that some tag has all the required attributes set, etc. and I don't want to special case any nodes. Also, I think nesting _is_ legal (ok maybe only in an `<embed>` but I'm not going to check the whole spec)



> `with dominate.document() as xxx: ...`
> should has no other effect then writing xxx.

This is explicitly asking for it to be the context. Not doing so would break the expectations of the API. If you don't want the effect, why not just write `xxx = dominate.document()`? 


If `xxx = ...` is still inside some outer context and causing problems:
> but as dominate doesnt support that I have to rewrite it from a top-down to a bottom-up manner.

That seems like an excessive reaction, why don't you just subclass `document` to do what you want?

```python
class mydoc(dominate.document):
    def _add_to_ctx(self): pass # don't add to contexts
```

With your example do what I think you want:

```shell
>>> with mydoc() as outer_doc:
...     with mydoc() as inner_doc:
...         p('meow')
...     print(inner_doc)
...
<dominate.tags.p at 1bb1a765160: 0 attributes, 1 child>
<!DOCTYPE html>
<html>
  <head>
    <title>Dominate</title>
  </head>
  <body>
    <p>meow</p>
  </body>
</html>
>>> print(outer_doc)
<!DOCTYPE html>
<html>
  <head>
    <title>Dominate</title>
  </head>
  <body></body>
</html>

```

**Comment by MarisaKirisame** [comment]
Thx. the subclass solution is excatly what I want.
